src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1131,143 +839,237 @@ PrimaryNodeForGroup(int32 groupId, bool *groupContainsNodes)
+	if (!MetadataSyncInNoConnectionMode(context))
+	{
+		List *updatedActivatedNodeList = NIL;
+
+		WorkerNode *node = NULL;
+		foreach_ptr(node, context->activatedWorkerNodeList)
 		{
-			bool localOnly = false;
-			DeleteAllReplicatedTablePlacementsFromNodeGroup(workerNode->groupId,
-															localOnly);
+			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_isactive,
+											BoolGetDatum(true));
+			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_metadatasynced,
+											BoolGetDatum(true));
+			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_hasmetadata,
+											BoolGetDatum(true));
+
+			updatedActivatedNodeList = lappend(updatedActivatedNodeList, node);
 		}
 
-		workerNode =
-			SetWorkerColumnLocalOnly(workerNode, Anum_pg_dist_node_isactive,
-									 BoolGetDatum(true));
+		/* reset activated nodes inside metadataSyncContext afer local update */
+		SetMetadataSyncNodesFromNodeList(context, updatedActivatedNodeList);
+	}
 
-		/* TODO: Once all tests will be enabled for MX, we can remove sync by default check */
-		bool syncMetadata = EnableMetadataSync && NodeIsPrimary(workerNode);
-		if (syncMetadata)
+	if (!localOnly)
+	{
+		WorkerNode *node = NULL;
+		foreach_ptr(node, context->activatedWorkerNodeList)
 		{
-			/*
-			 * We are going to sync the metadata anyway in this transaction, so do
-			 * not fail just because the current metadata is not synced.
-			 */
-			SetWorkerColumn(workerNode, Anum_pg_dist_node_metadatasynced,
-							BoolGetDatum(true));
+			SetNodeStateViaMetadataContext(context, node, BoolGetDatum(true));
+		}
+	}
+}
 
-			/*
-			 * Update local group id first, as object dependency logic requires to have
-			 * updated local group id.
-			 */
-			UpdateLocalGroupIdOnNode(workerNode);
 
-			nodeToSyncMetadata = lappend(nodeToSyncMetadata, workerNode);
-		}
+/*
+ * ActivateNodeList does some sanity checks and acquire Exclusive lock on pg_dist_node,
+ * and then activates the nodes inside given metadataSyncContext.
+ *
+ * The function operates in 3 different modes according to transactionMode inside
+ * metadataSyncContext.
+ *
+ * 1. MetadataSyncInNoConnectionMode(context):
+ *      Only collect commands instead of sending them to workers,
+ * 2. context.transactionMode == METADATA_SYNC_TRANSACTIONAL:
+ *      Send all commands using coordinated transaction,
+ * 3. context.transactionMode == METADATA_SYNC_NON_TRANSACTIONAL:
+ *      Send all commands using bare (no transaction block) connections.
+ */
+void
+ActivateNodeList(MetadataSyncContext *context)
+{
+	if (context->transactionMode == METADATA_SYNC_NON_TRANSACTIONAL &&
+		IsMultiStatementTransaction())
+	{
+		/*
+		 * prevent inside transaction block as we use bare connections which can
+		 * lead deadlock
+		 */
+		ereport(ERROR, (errmsg("do not sync metadata in transaction block "
+							   "when the sync mode is nontransactional"),
+						errhint("resync after SET citus.metadata_sync_transaction_mode "
```

Comment:
maybe change citus.metadata_sync_transaction_mode to citus.metadata_sync_mode as the parameters already implies transactions. if you agree, do not forget to change the error messages likee this